POO-Java/LearnToDoRealJava.md

```diff
@@ -55,43 +61,341 @@ public move(int x, int y) {
+        return (ArrayList<String>)
+            this.strings.clone();  // a clone still happens if this function is
+        }                          // called. We could, in practice, define an
+                                   // ArrayListView, but there's a much simpler
+                                   // solution (to be written soon TM).
+
+    public void clear_strings() {
+        this.strings.clear();
+    }
+}
+
+class Foo {
+    private Bar bar;
+
+    public Foo() {
+        this.bar = new Bar();
+    }
+
+    public BarView get_bar() {  // now returns an immutable view
+        return this.bar;  // implicit cast from Bar to BarView
+    }
+}
+
+// Takes a Bar and computes some result, but does not modify the Bar.
+// Since it doesn't modify the Bar, we can pass a BarView instead.
+void compute(BarView bar) { /* ... */ }
+
+var all_foos = new ArrayList<Foo>(/* ... */);
+for (var foo : all_foos) {
+    compute(foo.get_bar());  // no more unnecessary clones :)
+}
+```
+Conseguimos definir um método de garantir imutabilidade que:
+* não faz cópias desnecessárias, por isso
+* não precisa de lidar com exceções na criação das cópias;
+* não aceita modificações a `Foo.bar`.\
+Por exemplo, tentar chamar `Foo.get_bar().clear_strings()` dá um erro de
+compilação:\
+'`The method clear_strings() is undefined for the type BarView`', que faz todo o
+sentido, porque `BarView` não define esse método.
+
+### Nada disto seria necessário
+
+Embora a solução descrita esteja a resolver o problema, está a resolver um
+problema que *não devia existir*.\
+Em programação orientada a objetos, os objetos fazem o trabalho por nós, não o
+contrário. No exemplo anterior, estamos a "dissecar" um `Foo` através do método
+`get_bar()`, e a fazer o "trabalho" com a função `compute()`. A abordagem
+correta, em OOP, seria ter a função `compute()` como um método de instância de
+`Foo`:
+```java
+class Bar {
+    private ArrayList<String> strings;
+
+    public Bar() {
+        this.strings = new ArrayList<>();
+    }
+
+    public ArrayList<String> get_strings() {
+        return (ArrayList<String>) this.strings.clone();
+    }
+
+    public void clear_strings() {
+        this.strings.clear();
+    }
+}
+
+class Foo {
+    private Bar bar;
+
+    public Foo() {
+        this.bar = new Bar();
+    }
+
+    public void compute() { /* ... */ }
+}
+
+var all_foos = new ArrayList<Foo>(/* ... */);
+all_foos.forEach(Foo::compute);  // functional flavour
+```
+Isto evita a necessidade de clones, getters, interfaces, e em geral melhora a
+ergonomia do código. Conseguimos reduzir significativamente a quantidade de
+*[boilerplate code][BoilerplateCode]*. O código ficou mais sucinto, explícito, e
+legível.
 
-# Clones
+No entanto, isto nem sempre é possível. Pode acontecer que `Bar` e `Foo` façam
+parte de uma biblioteca externa que não possamos modificar. Neste caso, temos
+que recorrer a uma destas soluções, e escolher aquela que nos trará menos dor no
+futuro.
```

Comment:
Nenhuma das soluções que apresentaste podes fazer se `Foo` e `Bar` forem de um
external lib. Todas implicam extender a API destas. Mas it's fine, se não
quiseres dar reword disto acho que não vem grande mal ao mundo, realisticamente
libs externas que tem getters não fazem clone dos return values.

Reply:
O objetivo é instruir na criação de boas interfaces de programação, external libs é para lá da scope.

Reply:
De qualquer das formas, vou desambiguar o parágrafo.